Hydra CI failures

Hydra CI is suffering from a number of failures.
These failures should all be reproducible via `nox`.
For example, to reproduce the `lint-3.7` failure, run `nox -s lint-3.7` using python version 3.7.
To reproduce the `lint_plugins-3.8(hydra-nevergrad-sweeper)` failure, run `nox -s 'lint_plugins-3.8(hydra-nevergrad-sweeper)'` using python version 3.8.

```bash
git clone https://github.com/facebookresearch/hydra
cd hydra
# Check that we're using python version 3.7:
[[ $(python --version) =~ "Python 3.7." ]] || echo "not using python3.7"
pip install nox
nox -s lint-3.7
```

- [x] lint-3.7
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159774)
  - PR where CI failure occurred: https://github.com/facebookresearch/hydra/pull/2840
  - excerpt from log files:
		```
		  File "/home/circleci/project/.nox/lint-3-7/lib/python3.7/site-packages/flake8/plugins/manager.py", line 254, in _load_entrypoint_plugins
			eps = importlib_metadata.entry_points().get(self.namespace, ())
		AttributeError: 'EntryPoints' object has no attribute 'get'
		nox > [2024-01-18 12:31:05,224] Command flake8 --config .flake8 failed with exit code 1
		nox > [2024-01-18 12:31:05,225] Session lint-3.7 failed.
		```
  - Possible Remedies:
    - It might work to pin `importlib-metadata<=4; python_version <= '3.7'` in `hydra/requirements/dev.txt`.
	   Reference: https://stackoverflow.com/a/73932581
    - We could remove support for python3.7 (which has [reached end of life](https://devguide.python.org/versions/))
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] lint_plugins-3.8(hydra-nevergrad-sweeper)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: mypy
  - except from log files:
		```
		plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py:15: error: Skipping analyzing "nevergrad": module is installed, but missing library stubs or py.typed marker  [import-untyped]
		plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py:15: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
		plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py:79: error: Unused "type: ignore" comment  [unused-ignore]
		```
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] lint_plugins-3.8(hydra-rq-launcher)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: mypy
  - excerpt from log files:
		```
		plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:22: error: Unused "type: ignore" comment  [unused-ignore]
		plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:127: error: Argument after ** must be a mapping, not "Union[Dict[Union[str, bytes, int, Enum, float, bool], Any], Dict[Any, Any]]"  [arg-type]
		```
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] test_plugins-3.8(hydra-ax-sweeper)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: usage of pandas API
  - excerpt from log files:
		```
		  File "/home/circleci/project/.nox/test_plugins-3-8-hydra-ax-sweeper/lib/python3.8/site-packages/pandas/core/frame.py", line 675, in __init__
			raise ValueError("columns cannot be a set")
		ValueError: columns cannot be a set
		```
  - related PR: https://github.com/facebookresearch/hydra/pull/2777
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] test_plugins-3.8(hydra-optuna-sweeper)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: usage of old optuna version?
  - excerpt from log files:
		```
		../../.nox/test_plugins-3-8-hydra-optuna-sweeper/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:1899: in _warnings_warn
			warnings.warn(message, stacklevel=stacklevel + 1)
		E   sqlalchemy.exc.MovedIn20Warning: The ``declarative_base()`` function is now available as sqlalchemy.orm.declarative_base(). (deprecated since: 2.0) (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
		```
  - related PR: https://github.com/facebookresearch/hydra/pull/2360
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [ ] test_plugins-3.8(hydra-ray-launcher)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - related PR: https://github.com/facebookresearch/hydra/pull/2778
  - excerpt from log files:
		```
		  File "/home/circleci/project/.nox/test_plugins-3-8-hydra-ray-launcher/lib/python3.8/site-packages/ray/_private/services.py", line 2041, in determine_plasma_store_config
			raise ValueError(
		ValueError: Attempting to cap object store memory usage at 70864896 bytes, but the minimum allowed is 78643200 bytes.
		```
- [x] test_plugins-3.8(hydra-rq-launcher)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] test_plugins-3.8(hydra-submitit-launcher)
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: deprecation warning
  - excerpt from log files:
		```
		  File "/home/circleci/project/.nox/test_plugins-3-8-hydra-submitit-launcher/lib/python3.8/site-packages/pkg_resources/__init__.py", line 118, in <module>
			warnings.warn(
		DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
		```
	- possible remedies: update the pytest deprecation warnings filter used in `noxfile.py` or `pyproject.py`
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842
- [x] test_plugins_vs_core-3.8
  - [CI logs demonstrating failure](https://app.circleci.com/pipelines/github/facebookresearch/hydra/15515/workflows/5731cff6-06f1-45ee-814f-a17e14a04796/jobs/159776)
  - nature of failure: deprecation warning
  - excerpt from log files:
		```
		.nox/test_plugins_vs_core-3-8/lib/python3.8/site-packages/pkg_resources/__init__.py:118: in <module>
			warnings.warn(
		E   DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
		```
  - possible remedies: update the pytest deprecation warnings filter used in `noxfile.py` or `pyproject.py`
  - fixed by: https://github.com/facebookresearch/hydra/pull/2842

In this issue I've just focused on python3.7 and python3.8, which is where the most failures are occurring

--- a/.mypy.ini
+++ b/.mypy.ini
@@ -14,3 +14,6 @@ ignore_errors = True
 
 [mypy-nevergrad.*]
 ignore_missing_imports = True
+
+[mypy-rq.*]
+ignore_missing_imports = True

--- a/.mypy.ini
+++ b/.mypy.ini
@@ -11,3 +11,6 @@ ignore_missing_imports = True
 [mypy-hydra.grammar.gen.*]
 ignore_missing_imports = True
 ignore_errors = True
+
+[mypy-nevergrad.*]
+ignore_missing_imports = True

--- a/plugins/hydra_ax_sweeper/setup.py
+++ b/plugins/hydra_ax_sweeper/setup.py
@@ -30,7 +30,7 @@ setup(
         "ax-platform>=0.1.20,<0.2.1",  # https://github.com/facebookresearch/hydra/issues/1767
         "torch",
         "gpytorch<=1.8.1",  # avoid deprecation warnings. This can probably be removed when ax-platform is unpinned.
-        "pandas<1.5.0",  # https://github.com/facebook/Ax/issues/1153, unpin when upgrading ax
+        "pandas<1.5.0",  # https://github.com/facebook/Ax/issues/1153, unpin when upgrading ax to >=0.2.8
     ],
     include_package_data=True,
 )

--- a/plugins/hydra_ax_sweeper/setup.py
+++ b/plugins/hydra_ax_sweeper/setup.py
@@ -30,6 +30,7 @@ setup(
         "ax-platform>=0.1.20,<0.2.1",  # https://github.com/facebookresearch/hydra/issues/1767
         "torch",
         "gpytorch<=1.8.1",  # avoid deprecation warnings. This can probably be removed when ax-platform is unpinned.
+        "pandas<1.5.0",  # https://github.com/facebook/Ax/issues/1153, unpin when upgrading ax
     ],
     include_package_data=True,
 )

--- a/plugins/hydra_ax_sweeper/setup.py
+++ b/plugins/hydra_ax_sweeper/setup.py
@@ -31,6 +31,7 @@ setup(
         "torch",
         "gpytorch<=1.8.1",  # avoid deprecation warnings. This can probably be removed when ax-platform is unpinned.
         "pandas<1.5.0",  # https://github.com/facebook/Ax/issues/1153, unpin when upgrading ax to >=0.2.8
+        "numpy<1.25.0",  # same as above, silences deprecation from np.find_common_type for pandas <1.5
     ],
     include_package_data=True,
 )

--- a/plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py
+++ b/plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py
@@ -76,7 +76,7 @@ def create_nevergrad_parameter_from_override(override: Override) -> Any:
         if "log" in val.tags:
             scalar = ng.p.Log(lower=val.start, upper=val.end)
         else:
-            scalar = ng.p.Scalar(lower=val.start, upper=val.end)  # type: ignore
+            scalar = ng.p.Scalar(lower=val.start, upper=val.end)
         if isinstance(val.start, int):
             scalar.set_integer_casting()
         return scalar

--- a/plugins/hydra_optuna_sweeper/setup.py
+++ b/plugins/hydra_optuna_sweeper/setup.py
@@ -29,6 +29,7 @@ setup(
     install_requires=[
         "hydra-core>=1.1.0.dev7",
         "optuna>=2.10.0,<3.0.0",
+        "sqlalchemy~=1.3.0",  # TODO: Unpin when upgrading to optuna v3.0
     ],
     include_package_data=True,
 )

--- a/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_config.py
+++ b/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_config.py
@@ -3,11 +3,11 @@ import sys
 from dataclasses import dataclass, field
 from enum import Enum, IntEnum
 from importlib import import_module
+from importlib.metadata import version
 from typing import Any, Dict, List, Optional
 
 from hydra.core.config_store import ConfigStore
 from omegaconf import OmegaConf
-from pkg_resources import get_distribution
 
 
 @dataclass
@@ -178,11 +178,11 @@ def _pip_pkgs_default_factory() -> Dict[str, str]:
         "hydra_core": "${ray_pkg_version:hydra}",
         "ray": "${ray_pkg_version:ray}",
         "cloudpickle": "${ray_pkg_version:cloudpickle}",
-        "hydra_ray_launcher": get_distribution("hydra_ray_launcher").version,
+        "hydra_ray_launcher": version("hydra_ray_launcher"),
     }
 
     if sys.version_info < (3, 8):
-        d["pickle5"] = get_distribution("pickle5").version
+        d["pickle5"] = version("pickle5")
 
     return d
 

--- a/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_core.py
+++ b/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_core.py
@@ -32,6 +32,8 @@ def launch(
         f"sweep output dir: {sweep_dir}"
     )
 
+    # Avoid allocating too little memory in CI https://github.com/ray-project/ray/issues/11966#issuecomment-1318100747
+    launcher.ray_cfg.init.setdefault("object_store_memory", 78643200)
     start_ray(launcher.ray_cfg.init)
 
     runs = []

--- a/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py
+++ b/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py
@@ -130,7 +130,7 @@ def launch(
             sweep_config=sweep_config,
             task_function=launcher.task_function,
             singleton_state=singleton_state,
-            **enqueue_keywords,  # type: ignore
+            **enqueue_keywords,
         )
         jobs.append(job)
 

--- a/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py
+++ b/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py
@@ -19,7 +19,7 @@ from hydra.core.utils import (
 from hydra.types import HydraContext, TaskFunction
 from omegaconf import DictConfig, OmegaConf, open_dict
 from redis import Redis
-from rq import Queue  # type: ignore
+from rq import Queue
 
 from .rq_launcher import RQLauncher
 
@@ -130,7 +130,7 @@ def launch(
             sweep_config=sweep_config,
             task_function=launcher.task_function,
             singleton_state=singleton_state,
-            **enqueue_keywords,
+            **enqueue_keywords,  # type: ignore
         )
         jobs.append(job)
 

--- a/plugins/hydra_rq_launcher/setup.py
+++ b/plugins/hydra_rq_launcher/setup.py
@@ -29,7 +29,7 @@ setup(
         "cloudpickle",
         "fakeredis<1.7.4",  # https://github.com/dsoftwareinc/fakeredis-py/issues/3
         "hydra-core>=1.1.0.dev7",
-        "rq>=1.5.1",
+        "rq>=1.5.1,<1.12",
     ],
     include_package_data=True,
 )

--- a/plugins/hydra_submitit_launcher/tests/test_submitit_launcher.py
+++ b/plugins/hydra_submitit_launcher/tests/test_submitit_launcher.py
@@ -58,5 +58,6 @@ def test_example(tmpdir: Path) -> None:
             "-m",
             f"hydra.sweep.dir={tmpdir}",
             "hydra/launcher=submitit_local",
-        ]
+        ],
+        allow_warnings=True,
     )

--- a/pytest.ini
+++ b/pytest.ini
@@ -16,3 +16,5 @@ filterwarnings =
   ignore:.*Future Hydra versions will no longer change working directory at job runtime by default.*:UserWarning
   ; Jupyter notebook test on Windows yield warnings
   ignore:.*Proactor event loop does not implement add_reader family of methods required for zmq.*:RuntimeWarning
+  ; Ignore deprecation warning related to pkg_resources
+  ignore:.*pkg_resources is deprecated*

--- a/requirements/dev.txt
+++ b/requirements/dev.txt
@@ -5,6 +5,7 @@ build
 coverage
 flake8<6  # pinned due to https://github.com/savoirfairelinux/flake8-copyright/issues/19
 flake8-copyright
+importlib-metadata<5.0; python_version <= '3.7'
 isort==5.5.2
 mypy
 nox

